coding.go

```diff
@@ -2,44 +2,53 @@ package format
 
 import (
 	"fmt"
-	"sync"
-
 	blocks "github.com/ipfs/go-block-format"
 )
 
 // DecodeBlockFunc functions decode blocks into nodes.
 type DecodeBlockFunc func(block blocks.Block) (Node, error)
 
-type BlockDecoder interface {
-	Register(codec uint64, decoder DecodeBlockFunc)
-	Decode(blocks.Block) (Node, error)
-}
-type safeBlockDecoder struct {
-	// Can be replaced with an RCU if necessary.
-	lock     sync.RWMutex
+// Registry is a structure for storing mappings of multicodec IPLD codec numbers to DecodeBlockFunc functions.
+//
+// Registry includes no mutexing. If using Registry in a concurrent context, you must handle synchronization yourself.
+// (Typically, it is recommended to do initialization earlier in a program, before fanning out goroutines;
+// this avoids the need for mutexing overhead.)
+//
+// Multicodec indicator numbers are specified in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+// You should not use indicator numbers which are not specified in that table
+// (however, there is nothing in this implementation that will attempt to stop you, either).
+type Registry struct {
 	decoders map[uint64]DecodeBlockFunc
 }
 
+func (r *Registry) ensureInit() {
+	if r.decoders != nil {
+		return
+	}
+	r.decoders = make(map[uint64]DecodeBlockFunc)
+}
+
 // Register registers decoder for all blocks with the passed codec.
 //
 // This will silently replace any existing registered block decoders.
-func (d *safeBlockDecoder) Register(codec uint64, decoder DecodeBlockFunc) {
-	d.lock.Lock()
-	defer d.lock.Unlock()
-	d.decoders[codec] = decoder
+func (r *Registry) Register(codec uint64, decoder DecodeBlockFunc) {
+	r.ensureInit()
+	if decoder == nil {
+		panic("not sensible to attempt to register a nil function")
+	}
+	r.decoders[codec] = decoder
 }
 
-func (d *safeBlockDecoder) Decode(block blocks.Block) (Node, error) {
+func (r *Registry) Decode(block blocks.Block) (Node, error) {
 	// Short-circuit by cast if we already have a Node.
 	if node, ok := block.(Node); ok {
 		return node, nil
 	}
 
 	ty := block.Cid().Type()
-
-	d.lock.RLock()
-	decoder, ok := d.decoders[ty]
-	d.lock.RUnlock()
+	r.ensureInit()
```

Comment:
is there a way to avoid doing this work on every usage?

Reply:
I just copied this from the go-ipld-prime one https://github.com/ipld/go-ipld-prime/blob/master/multicodec/registry.go. But yeah happy to change it either here, or in a patch release if this ends up being annoying for people.